### ISSUE TITLE
ZooKeeper 3.6.4 quorum failing with Kafka 3.4.1

### ISSUE THREAD

**Issue by ppatierno** [strimzi/strimzi-kafka-operator#8675]
I am using the `main` branch and deploying the operator with `latest` image which includes the support for Kafka 3.4.1 (bringing a new ZooKeeper version 3.6.4).
With minikube (I tried with both an older 1.26.1 and newer 1.30.1 versions), the ZooKeeper ensamble gets stuck, the quorum is not formed, the operator is not able to connect to the ZooKeeper nodes so it cannot proceed with the Kafka pods creation.
The cluster deployment remains in this condition forever.

After long investigation it turned up that during the startup, one or more ZooKeeper pods, are not able to bind the quorum port 3888 to the Cluster IP but during the DNS resolution they get the loopback address instead.
Following a possible log at ZooKeeper startup where you can see the binding at 127.0.0.1:3888 instead of something like 172.17.0.4:3888 (so getting a valid not loopback IP address).

```shell
INFO 3 is accepting connections now, my election bind port: my-cluster-zookeeper-2.my-cluster-zookeeper-nodes.default.svc/127.0.0.1:3888 (org.apache.zookeeper.server.quorum.QuorumCnxManager) [ListenerHandler-my-cluster-zookeeper-2.my-cluster-zookeeper-nodes.default.svc/127.0.0.1:3888]
```

This way, the ZooKeeper node is not accessible from the others for forming a quorum.

It's also important to highlight I was not able to reproduce the issue on OpenShift or AKS instances and it's also not so easy to reproduce on Mac based machine (I am using Fedora).
Anyway the same exact minikube configuration works fine with previous Kafka 3.4.0 (bringing an older ZooKeeper version 3.6.3).

It seems the problem could be related to the newer ZooKeeper version 3.6.4 having a faster startup time when the DNS names registration is not already done at platform level (see minikube) so the name resolution provides the loopback address instead of the Cluster IP we want.

**Comment by ppatierno** [comment]
In order to address the problem we can think of three solutions.

### Solution 1

Changing the `server.X` configuration by using the 0.0.0.0 address for the local ZooKeeper node. It means binding the 2888 and 3888 to all network interfaces. So for example, on the ZooKeeper having ID = 1 using something like:

```shell
# Zookeeper nodes configuration
server.1=0.0.0.0:2888:3888:participant;127.0.0.1:12181
server.2=my-cluster-zookeeper-1.my-cluster-zookeeper-nodes.default.svc:2888:3888:participant;127.0.0.1:12181
server.3=my-cluster-zookeeper-2.my-cluster-zookeeper-nodes.default.svc:2888:3888:participant;127.0.0.1:12181
```

so the ZooKeeper binding on 3888 will be something like this:

```shell
INFO 1 is accepting connections now, my election bind port: /0.0.0.0:3888 (org.apache.zookeeper.server.quorum.QuorumCnxManager) [ListenerHandler-/0.0.0.0:3888]
```

I already tried this solution hacking the operator regarding the ZooKeeper configuration shell script.

### Solution 2

Achieving the same result as Solution 1 but using the `quorumListenOnAllIPs` parameter on ZooKeeper configuration.
NOTE: this solution doesn't seem to work. It would need more investigation because the behaviour should be the same as Solution 1.

### Solution 3

Having an init container, starting up before ZooKeeper, which checks the correct DNS names registration to be sure that the described issue cannot happen when ZooKeeper starts.
Not tried. It's worth only if someone thinks it's a good one.

I would be for Solution 1 or 2 (if it works) and maybe making it optional through a configuration parameter in the operator if someone has concern about binding on all interfaces. Otherwise I would keep this way as default.

**Comment by ppatierno** [comment]
We had more investigation on Solution 2 and the root cause is still a potential DNS resolution problem.
During a ZooKeeper leader election, when the reconfiguration is enabled, the so called "designated" leader is not the current one anymore, because its hostname is not resolved correctly (just <unresolved>), it's taken as a difference in the configuration and the reconfiguration restarts with a new leader election.
Anyway, the current leader seems to stick being the leader but it's not allowed to "commit" operations, so the ZooKeeper ensamble is stuck, the cluster operator is not able to connect and it cannot proceed with the Kafka brokers deployment.
This problem doesn't happen with the Solution 1.
We started a [thread](https://lists.apache.org/thread/vsff39wpsrxb431hdw9db4np12vytdpm) on the ZooKeeper users mailing list as well.

Following more details about the above issue.

Getting the log of the current ZooKeeper leader (ID=3) we see the following.
(Starting with ** you can see some additional logs added to `org.apache.zookeeper.server.quorum.Leader#getDesignatedLeader` in order to get more information.)

```shell
2023-06-19 12:32:51,990 INFO Have quorum of supporters, sids: [[1, 3],[1, 3]]; starting up and setting last processed zxid: 0x100000000 (org.apache.zookeeper.server.quorum.Leader) [QuorumPeer[myid=3](plain=127.0.0.1:12181)(secure=0.0.0.0:2181)]
2023-06-19 12:32:51,990 INFO ** newQVAcksetPair.getQuorumVerifier().getVotingMembers().get(self.getId()).addr = my-cluster-zookeeper-2.my-cluster-zookeeper-nodes.default.svc/172.17.0.6:2888 (org.apache.zookeeper.server.quorum.Leader) [QuorumPeer[myid=3](plain=127.0.0.1:12181)(secure=0.0.0.0:2181)]
2023-06-19 12:32:51,990 INFO ** self.getQuorumAddress() = my-cluster-zookeeper-2.my-cluster-zookeeper-nodes.default.svc/<unresolved>:2888 (org.apache.zookeeper.server.quorum.Leader) [QuorumPeer[myid=3](plain=127.0.0.1:12181)(secure=0.0.0.0:2181)]
2023-06-19 12:32:51,992 INFO ** qs.addr my-cluster-zookeeper-2.my-cluster-zookeeper-nodes.default.svc/172.17.0.6:2888, qs.electionAddr my-cluster-zookeeper-2.my-cluster-zookeeper-nodes.default.svc/172.17.0.6:3888, qs.clientAddr/127.0.0.1:12181 (org.apache.zookeeper.server.quorum.QuorumPeer) [QuorumPeer[myid=3](plain=127.0.0.1:12181)(secure=0.0.0.0:2181)]
2023-06-19 12:32:51,992 DEBUG zookeeper (org.apache.zookeeper.common.PathTrie) [QuorumPeer[myid=3](plain=127.0.0.1:12181)(secure=0.0.0.0:2181)]
2023-06-19 12:32:51,993 WARN Restarting Leader Election (org.apache.zookeeper.server.quorum.QuorumPeer) [QuorumPeer[myid=3](plain=127.0.0.1:12181)(secure=0.0.0.0:2181)]
```

So the leader is ZooKeeper with ID=3 and it was ACKed by the ZooKeeper node ID=1.
As you can see we are in the `Leader#startZkServer` method, and because of the reconfiguration enabled, the designatedLeader is processed. The problem is that the `Leader#getDesignatedLeader` is not returning “self” as leader but another one (ID=1), because of the difference in the quorum address.
From the above log, it’s not an actual difference in terms of addresses but the `self.getQuorumAddress()` is returning an <unresolved> one (even if it’s still the same hostname related to ZooKeeper-2 instance). This difference causes the allowedToCommit=false, meanwhile the ZooKeeper-2 is still reported as leader but it’s not able to commit, so prevents any requests and the ZooKeeper ensemble gets stuck.

```shell
2023-06-19 12:32:51,996 WARN Suggested leader: 1 (org.apache.zookeeper.server.quorum.QuorumPeer) [QuorumPeer[myid=3](plain=127.0.0.1:12181)(secure=0.0.0.0:2181)]
2023-06-19 12:32:51,996 WARN This leader is not the designated leader, it will be initialized with allowedToCommit = false (org.apache.zookeeper.server.quorum.Leader) [QuorumPeer[myid=3](plain=127.0.0.1:12181)(secure=0.0.0.0:2181)]
```

The overall issue could be related to DNS problems, with DNS records not registered yet during pod initialization (where ZooKeeper is running on Kubernetes). But we don’t understand why it’s not able to recover somehow.

Instead of using the quorumListenOnAllIPs=true we also tried a different approach by using the 0.0.0.0 address for the binding, so something like:

```
# Zookeeper nodes configuration
server.1=0.0.0.0:2888:3888:participant;127.0.0.1:12181
server.2=my-cluster-zookeeper-1.my-cluster-zookeeper-nodes.default.svc:2888:3888:participant;127.0.0.1:12181
server.3=my-cluster-zookeeper-2.my-cluster-zookeeper-nodes.default.svc:2888:3888:participant;127.0.0.1:12181
```

This way, the `self.getQuorumAddress()` is not suffering the same problem, it doesn’t return an <unresolved> address but always an actual one. No new leader election is needed and everything works fine.

**Comment by showuon** [comment]
@ppatierno , thanks for patiently investigating this issue. I agree the solution 1 (0.0.0.0 solution) is a good workaround for this slow DNS issue. But since 0.0.0.0 is binding to all network interface, if user has multiple network interface, it might expose unexpected security threat. Should we consider to make it optional (i.e. add a config to enable it, instead of default to use it)?

**Comment by ppatierno** [comment]
@showuon I am fine with the idea to make it optional, even because I wasn't able to reproduce the issue on environment like AKS or OpenShift (where the DNS registration seems to work much better). Anyway it could happen if DNS is slow, as it seems to happen on minikube.
We could think of an env var `STRIMZI_ZOOKEEPER_QUORUM_LISTEN_ON_ALL_IPS`, in the cluster operator Deployment, 
 which is `false` by default (if not defined), leaving the current configuration. If it's set to `true` the operator will use Solution 1 applied to the ZooKeeper configuration.

**Comment by scholzj** [comment]
You should not make it optional without a reason. Because it will make everything else a nightmare -> testing, upgrades, etc. So do you have a reason? This is also why I repeatedly suggested to open a PR to see how it works in different environments and not leave it for the last minute before the release without a chance to get any experience with it.

**Comment by ppatierno** [comment]
I have just opened the PR so that we can see how/if it works for different environments, then we can discuss about making it optional.
The reason was that some people raised security issue on using `0.0.0.0` as address to bind to on this Stack Overflow thread https://stackoverflow.com/questions/30940981/zookeeper-error-cannot-open-channel-to-x-at-election-address
Anyway as @showuon already checked, the described issue was fixed in 3.6.1 already https://issues.apache.org/jira/browse/ZOOKEEPER-2164 so we should be ok.

**Comment by ppatierno** [comment]
From running the tests via the PR #8702 I opened, it seems that the "Solution 1" doesn't work pretty well in case of having just one ZooKeeper node.
The issue is still the same and related to slow DNS names registration, but not able to recover.
Yesterday I also opened the [ZOOKEEPER-4708](https://issues.apache.org/jira/browse/ZOOKEEPER-4708) issue updated with the above problem as well.

At this point, it was worth exploring "Solution 3" about not starting ZooKeeper until it's possible to resolve the hostname on the pod. Instead of an init-container, I went with a few lines of bash script using `nslookup` tool added in the `zookeeper_run.sh`.

```shell
NODE=${ZOOKEEPER_ID}
echo "Resolving ${BASE_HOSTNAME}-$((NODE-1)).${BASE_FQDN} ..."
ipaddress=$(nslookup ${BASE_HOSTNAME}-$((NODE-1)).${BASE_FQDN} | grep "Address" | awk '{print $2}' | sed -n 2p)

while [ -z $ipaddress ]; do
  sleep 1
  echo "Resolving ${BASE_HOSTNAME}-$((NODE-1)).${BASE_FQDN} ..."
  ipaddress=$(nslookup ${BASE_HOSTNAME}-$((NODE-1)).${BASE_FQDN} | grep "Address" | awk '{print $2}' | sed -n 2p)
done

echo "IP address $ipaddress"
```

Running ZooKeeper pod with this addition showed the actual problem with minikube, by having 2 (out of 3) pods being able to resolve its own hostname later (after even 10-15 seconds).
Anyway, the solution seems to work (even with just one ZooKeeper node).
The usage of `nslookup` means the need for adding the package `bind-utils` in the base image.

Another option could be going with a simple Java application to run from the ZooKeeper run script and doing the checks instead of adding the `bind-utils` package.

Last but not the least, checking for DNS resolution ready adds some delay which could have influence on the probes.
It's also worth mentioning that on a system with really slow DNS resolution, the probes delay can be tweaked on the `Kafka` custom resource. (just saying that in my failing environment a.k.a. minikube, the default probes configuration was enough).

**Comment by ppatierno** [comment]
Taking the discussion offline, we agreed on going with the `nslookup` solution but allowing to disable it via a `STRIMZI_` env var on the cluster operator.
I will move forward to change the PR accordingly.

**Comment by nguyenthai0107** [comment]
Hello @ppatierno  
Could i know exactly version which implemented for resolved issue slow DNS query.
Thank you.

**Comment by ppatierno** [comment]
@nguyenthai0107 it should be from 0.36.1

**Comment by nguyenthai0107** [comment]
Hello @ppatierno .
Thanks for your reply, but look like still get stuck in with quorum failed
`2023-12-18 15:52:26,997 WARN Cannot open channel to 2 at election address A-1.NODE.NS.svc/x.x.x.x:3888 (org.apache.zookeeper.server.quorum.QuorumCnxManager) [QuorumConnectionThread-[myid=1]-5]
java.net.SocketTimeoutException: Connect timed out
        at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:551)
        at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:602)
        at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
        at java.base/java.net.Socket.connect(Socket.java:633)
        at java.base/sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:304)
        at org.apache.zookeeper.server.quorum.QuorumCnxManager.initiateConnection(QuorumCnxManager.java:383)
        at org.apache.zookeeper.server.quorum.QuorumCnxManager$QuorumConnectionReqThread.run(QuorumCnxManager.java:457)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
2023-12-18 15:52:26,999 WARN Cannot open channel to 3 at election address A-2.NODE.NS.svc/x.x.x.x:3888  tracing-kafka-(org.apache.zookeeper.server.quorum.QuorumCnxManager) [QuorumConnectionThread-[myid=1]-3]
java.net.SocketTimeoutException: Connect timed out
        at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:551)
        at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:602)
        at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
        at java.base/java.net.Socket.connect(Socket.java:633)
        at java.base/sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:304)
        at org.apache.zookeeper.server.quorum.QuorumCnxManager.initiateConnection(QuorumCnxManager.java:383)
        at org.apache.zookeeper.server.quorum.QuorumCnxManager$QuorumConnectionReqThread.run(QuorumCnxManager.java:457)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
2023-12-18 15:52:32,390 INFO Processing ruok command from /127.0.0.1:43190 (org.apache.zookeeper.server.NettyServerCnxn) [nioEventLoopGroup-4-1]
2023-12-18 15:52:32,440 INFO Processing ruok command from /127.0.0.1:43200 (org.apache.zookeeper.server.NettyServerCnxn) [nioEventLoopGroup-4-2]
2023-12-18 15:52:42,501 INFO Processing ruok command from /127.0.0.1:57504 (org.apache.zookeeper.server.NettyServerCnxn) [nioEventLoopGroup-4-1]`
Could you please take a look

**Comment by ppatierno** [comment]
Strimzi workaround is just a meetigation and sometimes could not work.
The future Kafka will come with the fixed ZooKeeper. The only think you can do is restarting ZK pods and see it that helps. There is no that much we can do. It's a big in ZK.

**Comment by avgojzack** [comment]
This Zookeeper ticket is also related, the name resolution doesn't heal itself it can get stuck in a tight failure loop forever:
https://issues.apache.org/jira/browse/ZOOKEEPER-4728

The problem is fixed in 3.8.4, but if you're running Kafka 3.6.2 in Strimzi .39x then you are currently stuck with it:
https://github.com/strimzi/strimzi-kafka-operator/blob/release-0.39.x/kafka-versions.yaml#L298
